File: src/Java/pipeline-core/src/main/java/com/pipeline/core/ActionTiming.java
package com.pipeline.core;

import java.util.Objects;

public record ActionTiming(
    StepPhase phase,
    int actionIndex,
    String actionName,
    long elapsedNanos,
    boolean success
) {
  public ActionTiming {
    phase = Objects.requireNonNull(phase, "phase");
    if (actionIndex < 0) throw new IllegalArgumentException("actionIndex must be >= 0");
    actionName = Objects.requireNonNull(actionName, "actionName");
    if (elapsedNanos < 0L) throw new IllegalArgumentException("elapsedNanos must be >= 0");
  }
}

File: src/CSharp/pipeline_services/core/ActionTiming.cs
using System;

namespace PipelineServices.Core;

public sealed class ActionTiming
{
    public ActionTiming(
        StepPhase phase,
        int actionIndex,
        string actionName,
        long elapsedNanos,
        bool success)
    {
        Phase = phase;
        if (actionIndex < 0)
        {
            throw new ArgumentOutOfRangeException(
                nameof(actionIndex),
                "actionIndex must be >= 0");
        }
        ActionIndex = actionIndex;
        ActionName = actionName ?? throw new ArgumentNullException(nameof(actionName));
        ElapsedNanos = elapsedNanos;
        Success = success;
    }

    public StepPhase Phase { get; }

    public int ActionIndex { get; }

    public string ActionName { get; }

    public long ElapsedNanos { get; }

    public bool Success { get; }

    [Obsolete("Use ActionIndex.")]
    public int Index => ActionIndex;
}
